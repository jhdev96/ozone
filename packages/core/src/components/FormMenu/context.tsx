import React from 'react';


interface ContextType {
  activeOption: {
    name: string,
    yPos: number,
    deltaY: number,
    menuSlider: React.MutableRefObject<HTMLDivElement | undefined>
  };
  setActiveOption: (params: ContextType['activeOption']) => void
};

export const MenuContext = React.createContext<[ContextType['activeOption'], ContextType['setActiveOption']]>([
  {
    name: '', 
    yPos: 7, 
    deltaY: 0, 
    menuSlider: React.createRef<HTMLDivElement | undefined>() as React.MutableRefObject<HTMLDivElement | undefined>
  },
  (_: ContextType['activeOption']) => {}
]);
export const MenuOptionContext = React.createContext<ContextType[]>([]);