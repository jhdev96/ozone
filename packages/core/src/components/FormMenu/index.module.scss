@import '../../scss/root';


.formMenu {
  position: relative;
  display: inline-flex;
  width: fit-content;
  height: 100%;
  min-width: 150px;
  padding: 10px 0;
}

.sliderWrapper {
  min-width: 8px;

  .slider {
    width: 4px;
    height: 22px;
    margin-top: 5px;
    border-radius: 10px;
    background-color: var(--blue-core);
    transition: 0.4s;
  }
}

.item {
  display: block;
  padding: 10px 20px;
  text-decoration: none !important;
  cursor: pointer;
}

.light {
  background-color: var(--gray-core);
  color: var(--black-core);

  .item:hover {
    color: #000;
  }
}

.dark {
  background-color: var(--black-3);
  color: var(--gray-core);

  .item:hover {
    color: #fff;
  }
}

.active { color: var(--blue-core) !important; }

.transition {
  height: 8px !important;
  width: 8px !important;
  border-radius: 50%;
}