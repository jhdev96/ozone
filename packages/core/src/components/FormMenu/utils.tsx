import React from 'react';
import styles from './index.module.scss';


export const createTransition = (
  slider: React.MutableRefObject<HTMLDivElement | undefined>, 
  delta: number
) => {
  /** Add a transition to the menu slider when moving to an option
   * that is more than 50 pixels away from the current option.
   * 
   * @param slider // a reference to the menu slider element
   * @param delta // the difference between the two options
   */
  if (Math.abs(delta) > 50) {
    slider.current!.classList.add(styles.transition);

    setTimeout(() => {
      slider.current!.classList.remove(styles.transition)
    }, 400);
  }
}