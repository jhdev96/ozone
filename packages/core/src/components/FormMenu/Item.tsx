import { v4 } from 'uuid';
import React, { useContext, useEffect } from 'react';
import { MenuContext } from './context';
import { createTransition } from './utils';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export interface ItemProps {
  /** 
   * The unique identifier for the item. 
   */
  id?: string;
  /** 
   * Callback to be invoked when an item is clicked. 
   */
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
  /** 
   * The name of the item. 
   */
  children?: string;
}

const Item = React.forwardRef<HTMLDivElement, ItemProps>(
  function Item({
    id        = v4(),
    onClick   = () => {},
    children}: ItemProps,
    ref: React.ForwardedRef<HTMLDivElement>): JSX.Element 
  {
    const [activeOption, setActiveOption] = useContext(MenuContext);
    const isActive = activeOption.name === children;

    const makeActive = () => {
      /** 
       * Make the menu option active. This involves
       * updating context data and setting a transition
       * for the menu slider.
       */
      const htmlElem = document.getElementById(id);

      if (!htmlElem) return;

      const dy = activeOption.yPos - htmlElem.offsetTop - 2;

      if (dy !== 0) {
        createTransition(activeOption.menuSlider, dy);
        setActiveOption({
          ...activeOption,
          name: htmlElem.textContent || '', 
          yPos: htmlElem.offsetTop - 2,
          deltaY: dy
        });
      }
    }

    const handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
      makeActive();
      onClick(e);
    }

    useEffect(() => {
      isActive && makeActive();
    }, [isActive])

    return (
      <div
        onClick={handleClick}
        id={id}
        className={classNames(
          styles.item,
          isActive && styles.action
        )}
        ref={ref}
      >
        {children}
      </div>
    )
  }
)

export default Item;

Item.displayName = 'FormMenuItem';