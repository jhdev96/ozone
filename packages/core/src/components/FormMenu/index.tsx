import React, { useState, useEffect, LegacyRef } from 'react';
import { MenuContext } from './context';
import { ThemeType } from '../../utils/theme';
import Item from './Item';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


interface FormMenuProps {
  /** 
   * Modifies the menu's visual appearance for 
   * light and dark themes. 
   */
  theme?: ThemeType;
  /** 
   * Sets the active item. 
   */
  activeItem: string;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The menu items to be rendered inside the form menu. 
   */
  children?: React.ReactNode;
}

export default function FormMenu({
  theme = "light",
  activeItem,
  dataTestId,
  children
}: FormMenuProps): JSX.Element 
{
  const menuSlider = React.useRef<HTMLDivElement>();
  const STARTING_Y_POS = 7;
  const [active, setActive] = useState({
    name: activeItem, 
    yPos: STARTING_Y_POS,
    deltaY: 0,
    menuSlider
  });

  useEffect(() => {
    setActive({...active, name: activeItem});
  }, [activeItem]);

  return (
    <div 
      className={classNames(
        styles.formMenu, 
        styles[theme]
      )}
      data-testid={dataTestId}
    >
      <MenuContext.Provider value={[active, setActive]}>
        <div className={styles.sliderWrapper}>
          <div 
            className={classNames(styles.slider)} 
            style={{marginTop: active.yPos}} 
            ref={menuSlider as LegacyRef<HTMLDivElement>}
          />
        </div>
        <div className={styles.items}>{children}</div>
      </MenuContext.Provider>
    </div>
  )
}

FormMenu.Item = Item;
FormMenu.displayName = 'FormMenu';