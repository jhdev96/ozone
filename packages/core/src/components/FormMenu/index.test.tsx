import React from "react";
import { render, fireEvent, screen } from '@testing-library/react';
import FormMenu from ".";


describe('<FormMenu />', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('renders a form menu with 3 items', () => {
    render(
      <FormMenu activeItem="A" dataTestId="test">
        <FormMenu.Item id="a">A</FormMenu.Item>
        <FormMenu.Item id="b">B</FormMenu.Item>
        <FormMenu.Item id="c">C</FormMenu.Item>
      </FormMenu>
    );

    expect(document.querySelector('.items')!.children)
      .toHaveLength(3);

    expect(screen.getByTestId("test")).toMatchSnapshot();
  });

  it('calls `onClick` when an item is clicked', () => {
    const onClick = jest.fn();

    render(
      <FormMenu activeItem="A" dataTestId="test">
        <FormMenu.Item>A</FormMenu.Item>
        <FormMenu.Item 
          id="b" 
          onClick={onClick}
        >
          B
        </FormMenu.Item>
        <FormMenu.Item>C</FormMenu.Item>
      </FormMenu>
    );

    fireEvent.click(document.querySelector('#b')!);
    expect(onClick).toBeCalled;
  });
});