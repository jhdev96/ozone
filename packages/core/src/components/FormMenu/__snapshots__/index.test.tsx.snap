// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormMenu /> renders a form menu with 3 items 1`] = `
<div
  class="formMenu light"
  data-testid="test"
>
  <div
    class="sliderWrapper"
  >
    <div
      class="slider"
      style="margin-top: 7px;"
    />
  </div>
  <div
    class="items"
  >
    <div
      class="item action"
      id="a"
    >
      A
    </div>
    <div
      class="item"
      id="b"
    >
      B
    </div>
    <div
      class="item"
      id="c"
    >
      C
    </div>
  </div>
</div>
`;
